experimental/ast/slice.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ast
+
+// Slice is a type that offers the same interface as an ordinary Go
+// slice.
+//
+// This is used to provide a consistent interface to various AST nodes that
+// contain a variable number of "something", but the actual backing array
+// is some compressed representation.
+type Slice[T any] interface {
+	// Len returns this slice's length.
+	Len() int
+
+	// At returns the nth value of this slice.
+	//
+	// Panics if n >= Len().
+	At(n int) T
+
+	// Iter is an iterator over the slice.
+	Iter(yield func(int, T) bool)
```

Comment:
Should we be *returning* the function? Seems like unnecessary overhead to me... ugh. I really don't want to have to write `slice.All()(func()...)` in the legalizer lol.

I'll change it to this if you insist but I'm just double-checking it's what you want.

Reply:
Oh, and we have `internal/iter` now so it's a bit harder to justify not doing this... ughhghg.

Reply:
I'm not a huge fan that it must return the function, but that is the expected pattern. Otherwise, it's not obvious from the type signature that this is intended for use with range-over-func. I'm pretty sure they suggest this convention so you could also add alternate iteration methods that accept other parameters, yet still have a uniform look to the signatures.